### DISCUSSION TITLE
Issue in writing out netcdf file with a large domain

### DISCUSSION THREAD

**Discussion by tais958** [NASA-LIS/LISF#1251] [Q&A] [answered]
Hi, I tried to run LIS (noah-MP 4.0.1 retrospective simulation) with a relatively large domain covering a large portion of the CONUS (1-km grid spacing, 2751 x 1891 grid points). It stopped at the point when it tried to write the first netcdf output file, so I suspect it is due to the issue with a large grid size. Is anyone has any experience on handling large output file in netcdf? Any comments are much appreciated!

**Comment by dmocko** [comment]
Hi @tais958 

Yes, I regularly run a domain with 2560 x 1920 grid points, and write data to netCDF.

When the job fails when writing the first LIS_HIST file, it often is from an issue with the MODEL OUTPUT TBL file.

Please check the end of ALL of your lislog files.  For example, try "tail lislog.*" and look for any error messages
at the end of the log files.

If you don't see anything, you can point me to your directory on Discover, or attach your log files and job output.

**Reply by tais958** [reply]
Thanks @dmocko! I asked because I've been running smoothly with the same configuration except the domain size was much smaller. I wonder what might be causing the job fails in the MODEL OUTPUT TBL file? Please find the attached OUTPUT TBL file as well as the logest lislog file. Thank you for your help!

[NOAH36_OUTPUT_LIST.TBL.txt](https://github.com/NASA-LIS/LISF/files/10280699/NOAH36_OUTPUT_LIST.TBL.txt)
[lislog_ol.0001.txt](https://github.com/NASA-LIS/LISF/files/10280700/lislog_ol.0001.txt)

**Reply by dmocko** [reply]
Thanks.  I don't see anything in either file that might cause an issue.  But check the end of all your lislog files to be sure.  Is there any error message from the job submission?  It's rare for the job to just stop without at least some message.

**Reply by tais958** [reply]
I saw there are some related to MPI in lis.err as attached here, but was not able to identify specific error. Any suggestions?

[lis.err.txt](https://github.com/NASA-LIS/LISF/files/10280775/lis.err.txt)

**Reply by dmocko** [reply]
Thanks for sharing.  These lines caught my eye (e.g.):

```
Fatal error in PMPI_Bcast: Other MPI error, error stack:
Fatal error in PMPI_Reduce: Internal MPI error!, error stack:
Fatal error in PMPI_Gatherv: Internal MPI error!, error stack:
```

It may be that you are out of memory.  But please be sure that you set the stack size to unlimited in your job script:

```
ulimit -s unlimited
mpirun -np 560 ./LIS -f lis.config
```

**Reply by tais958** [reply]
Thanks! I have set the stack size as unlimited in my job script. I see you have cpu number 560 here, is this the exact number you used for your large domain run? If true, I might have set the cpu number too small (less than 200). Would it be possible the cause?

**Reply by dmocko** [reply]
I really can't say, because I don't know the hardware specifications (e.g., amount of memory per node) of the system on which you are running.

On another system on which I run, which has less memory per node, I specify in my job script that I only want to run using a portion of the procs on each node.  On Discover, you specify this using "--ntasks-per-core", but I don't know about the system on which you are trying to run.  But if use only a portion of procs on each node, then you may have available memory to run.  For example, if you have 128 procs on each node, you can try to use only 64 procs to run on each node.  That way, only 64 procs are fighting for the total memory on the node, instead of all 128.

**Reply by tais958** [reply]
Hello @dmocko . At one point, I was able to get around this by using GDAS analysis as the lowest level of forcing dat and overlaid NLDAS2 on top of it. Did not know why but it worked. Now I wanted to dig into this more. Wondering any chance if you have any guess on the cause of this? Thanks.

**Reply by dmocko** [reply]
It probably worked because you have one or more 1-km grid points that are outside of the NLDAS-2 domain.  The NLDAS-2 domain does not include forcing over the ocean, while GDAS forcing does.  So when you use GDAS with NLDAS-2 overlaid on top of it, it will use GDAS for those forcing points outside of the NLDAS-2 domain, as GDAS had forcing data globally over both land and ocean.

Please look at all of your lislog files when trying to run with just NLDAS-2.  There should be an error message at the end of one or more of them that will list the x/y point that does not have NLDAS-2 forcing.

**Reply by tais958** [reply]
Thanks @dmocko. I suspected the issue as you mentioned, but I wasn't able to see relevant error messages in any of the log files. Is there an option that I can easily switch on the debug mode? I would like to see where exactly it stopped as the log files were not giving the info. Thank you.

**Reply by tais958** [reply]
I also wondered why it has always stopped when in the process of output model variables but not during the model integration. If there's no forcing data for any of the grids, presumably it should stop progress when in forward integration. Any comments?

**Reply by dmocko** [reply]
cd LISF/lis/make, type "gmake realclean", and then when you configure, choose this option:
`Optimization level (-3=strict checks with warnings, -2=strict checks, -1=debug, 0,1,2,3, default=2): -1`
Enter "-1" for the above option, and then configure and compile as you did before.

**Reply by dmocko** [reply]
If you are on Discover, please point me to your lis.config and lis_input file.  If not, please attach them (along with your MODEL OUTPUT TBL file and forcing_variables.txt file).

**Reply by tais958** [reply]
> If you are on Discover, please point me to your lis.config and lis_input file. If not, please attach them (along with your MODEL OUTPUT TBL file and forcing_variables.txt file).

Thanks @dmocko. Here is a zip file which contains requested files you mentioned. Please let me know if I missed anything.
[LIS_files.zip](https://github.com/user-attachments/files/16664746/LIS_files.zip)

**Reply by dmocko** [reply]
HI @tais958 

You missed the end of your lislog_ol.0005 file.  It reads:

```
 undefined value found for forcing variable tair in NoahMP401
 for tile          114 latitude =    29.07000     longitude =   -90.52002    
 [ERR] endrun is being called
```

That lat/lon grid point does not have NLDAS-2 forcing data.  You will need to exclude it from your simulation using LDT,
or use GDAS and overlay NLDAS-2 data, as you did before.

Also, you didn't include this file.  Can you please attach it?

`LIS domain and parameter data file:    ../test1/lis_input.nldas.noahmp401.d01_1km.nc`

**Reply by tais958** [reply]
Thank you @dmocko. I thought it would appear in the largest size log file. Then I know what really caused the termination. Sorry for not providing the LIS domain and parameter data file. But it says the file size was too big to upload. Could you try accessing it through this google drive [link:](https://drive.google.com/file/d/1Mx_CpoFlIoQABDNDjza2nBBI7pDACkmg/view?usp=drive_link) ?

**Reply by dmocko** [reply, marked as answer]
Hi @tais958 

Thanks, I received your lis_input file.  At this point, you have two choices:

1. Use GDAS or some other forcing datasets as background, overlaying NLDAS-2, so any points outside of the NLDAS domain will use GDAS or some other forcing instead.
2. Mask out any land points not in the NLDAS domain (so they won't run and won't produce any output).

To do the latter, run LIS for a single timestep, outputting just one Forcing variable (such as Tair_F) in your MODEL OUTPUT file.

Also, use template LSM instead of Noah-MP for this short run.  For example, in your lis.config:

```
Surface model output interval:     15mn
Land surface model:    none
TEMPLATE model timestep:  15mn
```

Make the Ending time just 15 minutes later than the Starting time in your lis.config file.

You can output this LIS_HIST file in any format that you want.  But you will need to convert it into a real*4 binary file, where every point with valid forcing data has a value of "1.0", and all other points have a value of "0.0".  The file size should be equal to: 2751 X 1891 X 4.

Then, re-run LDT, changing these settings:

```
#Landmask parameter inputs
Create or readin landmask:              "readin"
Landmask data source:                   [same as what you used before]
Landmask file:                          ["Your real*4 binary file from above".1gd4r] - 1gd4r is an optional file suffix indicating real*4
Landmask spatial transform:             none
Landmask fill option:                   none
Landmask map projection:                latlon
Landmask lower left lat:                  [these next 6 values should be the same as in your "Run domain" values]
Landmask lower left lon:                
Landmask upper right lat:                 
Landmask upper right lon:                
Landmask resolution (dx):                 
Landmask resolution (dy):                  
```
LDT will then produce a lis_input file with a LANDMASK field that only has valid NLDAS-2 forcing for all grid points.  You can change your lis.config file back to what you had been using with Noah-MP-4.0.1 LSM, and the model should run as expected.

One thing to watch out for is that the big_endian/little_endian flag you chose when you configured LDT must match the type of endian-ness of the real*4 binary file when you created it.

**Reply by tais958** [reply]
Thank you so much for the instructions, @dmocko! This is extremely helpful. I'll try generating the landmask file and see how it goes.

**Reply by tais958** [reply]
Hi @dmocko, sorry for another question here. I followed the steps and created a new lis_iput file (https://drive.google.com/file/d/1AdtbdqDSZr_lyRpY4u_H8RLDlC2BLb6b/view?usp=drive_link). However, it stopped when reading the elevation as it showed in the lis.log:

[INFO] Reading mask from ../test1/lis_input.nldas.noahmp401.d01.mask.new.nc
 [INFO] Saving global land mask...
 [INFO] Reading landcover map from ../test1/lis_input.nldas.noahmp401.d01.mask.new.nc
 [INFO] Reading surfacetype map from ../test1/lis_input.nldas.noahmp401.d01.mask.new.nc
 [INFO] Finished reading landmask and landcover data
 [INFO] Reading elevation map from ../test1/lis_input.nldas.noahmp401.d01.mask.new.nc
 [ERR] Error in nf90_inq_dimid in read_elevation Stopping.
 [ERR] endrun is being called

I did not assign the "1" and "0" values for the grid points of the single variable in binary file. Instead, I modified the LANDMASK and DOMAINMASK afterward before running LIS. Not sure if this might lead to the error. Do you have any suggestion? Thanks!

**Reply by dmocko** [reply]
Please attach your ldt.config file

**Reply by tais958** [reply]
Please see the attached ldt.config file. Thanks!
[ldt.config.noahmp401_params.1km.zip](https://github.com/user-attachments/files/16753048/ldt.config.noahmp401_params.1km.zip)

**Reply by tais958** [reply]
Hi @dmocko.  After looking through the differences between the original and processed (assigned LANDMASK values using nco) netcdf files, I found the error was caused by incomplete dimension list which got lost after processing. It worked after adding these dimensions back to the processed netcdf file:

elevbins = 1 ;
slopebins = 1 ;
aspectbins = 1 ; 

Thanks for your helps!
